File: xcrop/orchestrator/app/homie_client.py
import httpx

from app.config import Settings

# Mirrors ../../src/xGIS.AddIn/Agent/ClaudeAgentService.cs's reason for existing: the
# gateway (backend/app/routes/chat.py) holds the only real Anthropic key and meters usage
# against the signed-in account's credit balance - xcrop must never call Anthropic
# directly, or AI usage here would be unbilled and unaudited against the same account the
# rest of the Homie platform tracks.


class HomieApiError(Exception):
    def __init__(self, status_code: int, detail: str):
        self.status_code = status_code
        self.detail = detail
        super().__init__(f"Homie API error {status_code}: {detail}")


class HomieClient:
    def __init__(self, settings: Settings):
        if not settings.homie_api_key:
            raise HomieApiError(401, "No Homie API key configured. Set one in xcrop Settings.")
        self._base_url = settings.homie_api_base.rstrip("/")
        self._headers = {"Authorization": f"Bearer {settings.homie_api_key}"}

    async def whoami(self) -> dict:
        return await self._get("/v1/me")

    async def chat(
        self,
        messages: list[dict],
        system: str | None = None,
        max_tokens: int = 1024,
    ) -> dict:
        """Calls the same /v1/chat gateway the ArcGIS Pro Add-in uses - request/response
        shape mirrors Anthropic's Messages API (see backend/app/routes/chat.py)."""
        body = {"messages": messages, "system": system, "max_tokens": max_tokens}
        return await self._post("/v1/chat", body)

    async def _get(self, path: str) -> dict:
        try:
            async with httpx.AsyncClient(timeout=30) as client:
                response = await client.get(f"{self._base_url}{path}", headers=self._headers)
        except httpx.HTTPError as exc:
            raise HomieApiError(503, f"Couldn't reach the Homie backend at {self._base_url}: {exc}") from exc
        return self._parse(response)

    async def _post(self, path: str, body: dict) -> dict:
        try:
            async with httpx.AsyncClient(timeout=60) as client:
                response = await client.post(f"{self._base_url}{path}", headers=self._headers, json=body)
        except httpx.HTTPError as exc:
            raise HomieApiError(503, f"Couldn't reach the Homie backend at {self._base_url}: {exc}") from exc
        return self._parse(response)

    def _parse(self, response: httpx.Response) -> dict:
        if response.status_code >= 400:
            detail = response.text
            try:
                detail = response.json().get("detail", detail)
            except ValueError:
                pass
            raise HomieApiError(response.status_code, detail)
        return response.json()

File: homie-presentation/lib/homieClient.ts
// Mirrors xcrop/orchestrator/app/homie_client.py and src/xGIS.AddIn/Agent/ClaudeAgentService.cs's
// reason for existing: backend/app/routes/chat.py's gateway holds the only real Anthropic
// key and meters usage against the signed-in account's credit balance - this app must
// never call Anthropic directly, or AI usage here would be unbilled and unaudited against
// the same account the rest of the Homie platform tracks. The API key itself is supplied
// by the browser on every request (see app/page.tsx) and only ever forwarded server-side
// to this same Homie gateway - it's never persisted here.

const HOMIE_API_BASE = process.env.HOMIE_API_BASE || "https://homie-platform.onrender.com";

export class HomieApiError extends Error {
  constructor(
    public status: number,
    detail: string,
  ) {
    super(detail);
  }
}

interface AnthropicContentBlock {
  type: string;
  text?: string;
}

interface ChatResponse {
  message: { content: AnthropicContentBlock[] };
  deducted_credits: number;
  remaining_credits: number;
}

/** Calls the same /v1/chat gateway xcrop and Homie GIS use - request/response shape
 * mirrors Anthropic's Messages API (see backend/app/routes/chat.py). Returns the
 * concatenated text of every text content block, since this app only ever asks for a
 * single structured-JSON text reply, never tool use. */
export async function homieChat(apiKey: string, system: string, userMessage: string): Promise<ChatResponse> {
  let response: Response;
  try {
    response = await fetch(`${HOMIE_API_BASE}/v1/chat`, {
      method: "POST",
      headers: { "Content-Type": "application/json", Authorization: `Bearer ${apiKey}` },
      body: JSON.stringify({
        system,
        messages: [{ role: "user", content: userMessage }],
        max_tokens: 4096,
      }),
    });
  } catch (err) {
    throw new HomieApiError(503, `Couldn't reach the Homie backend at ${HOMIE_API_BASE}: ${String(err)}`);
  }

  if (!response.ok) {
    let detail = await response.text();
    try {
      detail = JSON.parse(detail).detail ?? detail;
    } catch {
      // Non-JSON error body - use the raw text as-is.
    }
    throw new HomieApiError(response.status, detail);
  }

  return response.json();
}

export function extractText(response: ChatResponse): string {
  return response.message.content
    .filter((block) => block.type === "text" && block.text)
    .map((block) => block.text)
    .join("\n");
}
